apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/InstanceHostManager.cs

```diff
@@ -0,0 +1,96 @@
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// host manager for spanner data client to get and cache the endpointUris for an instance.
+    /// </summary>
+    public sealed class InstanceHostManager
+    {
+        private static ConcurrentDictionary<string, RepeatedField<string>> _instanceEndpointMapping =
+            new ConcurrentDictionary<string, RepeatedField<string>>();
+        Random random = new Random();
+        private bool IsResourceBasedRoutingEnabled
+        {
+            get
+            {
+                string resourceBaseRoutingFlagValue = Environment.GetEnvironmentVariable("GOOGLE_CLOUD_ENABLE_RESOURCE_BASED_ROUTING");
+                return StringComparer.InvariantCultureIgnoreCase.Equals(resourceBaseRoutingFlagValue, "true");
+            }
+        }
+
+        static InstanceHostManager() { }
+
+        /// <summary>
+        /// The default instance host manager
+        /// </summary>
+        public static InstanceHostManager Default => new InstanceHostManager();
+
+        /// <summary>
+        /// get host for specific instance
+        /// </summary>
+        /// <param name="projectId"></param>
+        /// <param name="instanceId"></param>
+        /// <returns>host</returns>
+        public string GetHost(string projectId, string instanceId)
+        {
+            if (IsResourceBasedRoutingEnabled)
+            {
+                string cacheKey = $"{projectId}-{instanceId}";
+                if (!_instanceEndpointMapping.ContainsKey(cacheKey))
+                {
+                    RepeatedField<string> instanceEndpointUris = GetInstanceEndpointUriAsync(projectId, instanceId).Result;
+                    _instanceEndpointMapping.GetOrAdd(cacheKey, instanceEndpointUris);
+                }
+                return _instanceEndpointMapping[cacheKey].ElementAtOrDefault(random.Next(_instanceEndpointMapping[cacheKey].Count));
+            }
+            return null;
+        }
+
+        private async Task<RepeatedField<string>> GetInstanceEndpointUriAsync(string projectId, string instanceId)
+        {
+            InstanceAdminClient instanceAdminClient =
+                    await InstanceAdminClient.CreateAsync().ConfigureAwait(false);
+            // Create request for instance with endpoints.
+            GetInstanceRequest request = new GetInstanceRequest
+            {
+                InstanceName = new InstanceName(projectId, instanceId),
+                FieldMask = new FieldMask { Paths = { "endpoint_uris" } }
+            };
+            try
+            {
+                // Get list of available endpointUris from GetInstance.
+                Instance instance = instanceAdminClient.GetInstance(request);
+                return instance.EndpointUris;
+            }
+            catch (RpcException gRpcException) when (gRpcException.StatusCode == StatusCode.PermissionDenied)
```

Comment:
@skuruppu 
WDYT if exception other than `PermissionDenied` has been thrown while retrieving the `endpointUris`, Should it continue with default host or throws exception to user?

Reply:
We had the following requirements in the spec:

**GetInstance permissions error**
This means the user is likely running under a custom role that doesn't have spanner.instances.get permissions. Fail with a warning message encouraging the user to add the permission and revert back to using the endpoint configured in the input options.

"The client library attempted to connect to an endpoint closer to your Cloud Spanner data but was unable to do so. The client library will fallback and route requests to the endpoint given in the client options, which may result in increased latency. We recommend including the scope https://www.googleapis.com/auth/spanner.admin so that the client library can get an instance-specific endpoint and efficiently route requests."

**other error**
Retry the request with fuzzed exponential backoff. Hard failure for repeated error responses.

**0 endpoints returned**
There is no instance config override. Subsequent request should use the endpoint that was used in the GetInstance request (either the global default or user-specified override).

**1+ endpoint**
Subsequent requests should use the first returned endpoint.

We expect users to run into the PERMISSION_DENIED error because we suspect lots of people run their prod applications without admin permissions. That's why it's handled specifically.

But beyond that, we expect GAX to retry requests. If the retries fail, the backend team decided that it should be a hard failure. I'm not sure what other types of errors there could be such that we should revert back to the default service address.

Reply:
OK, I would personally still ignore almost all errors, even after retries, but if it is by design I'm fine with it.
Same with always returnning the first host if more than one are returned, still a little weird, but if by design then it's allright.